File: tictactoe.py
"""
Tic Tac Toe Player
"""

import math
from copy import deepcopy
from random import choice, shuffle

X = "X"
O = "O"
EMPTY = None


def initial_state():
    """
    Returns starting state of the board.
    """
    return [[EMPTY, EMPTY, EMPTY],
            [EMPTY, EMPTY, EMPTY],
            [EMPTY, EMPTY, EMPTY]]


def player(board):
    """
    Returns player who has the next turn on a board.
    """
    player_X_turns ,player_O_turns = 0, 0
    for row in board:
        for tile in row:
            if tile == X:
                player_X_turns += 1
            elif tile == O:
                player_O_turns += 1
    
    if player_X_turns > player_O_turns:
        return O
    else:
        return X


def actions(board):
    """
    Returns set of all possible actions (i, j) available on the board.
    """
    # init moves set
    moves = set()
    # find len of row & col in board
    row, col = len(board), len(board[0])
    # find & add moves
    for i in range(row):
        for j in range(col):
            if board[i][j] == EMPTY:
                moves.add((i, j))

    return list(moves)


def result(board, action):
    """
    Returns the board that results from making move (i, j) on the board.
    """
    # copy board
    board_copy = deepcopy(board)
    # find which pllayer's turn is in the board
    turn = player(board)
    # try & make the move if possible on the copied board else raise exception
    i, j = action
    try:
        board_copy[i][j] = turn
    except IndexError:
        raise IndexError("Invalid move")
    
    return board_copy


def winner(board):
    """
    Returns the winner of the game, if there is one.
    """
    # bond dims
    board_len = len(board)
    # go through board and find: 
    #   row sequence, col_sequence, diagonal sequence, cross diagonal sequence
    #   if any
    diag, cross_diag = set(), set()
    for i in range(board_len):
        # reset the row, col sequences
        row_seq, col_seq = set(), set()
        for j in range(board_len):
            row_seq.add(board[i][j])
            col_seq.add(board[j][i])
            # find and add diagonal element in our set
            if i == j:
                diag.add(board[i][j])
                cross_diag.add(board[i][board_len-j-1])
        # return winner if any sequence of player is found(exclude empty sequence)
        if len(row_seq) == 1 and not EMPTY in row_seq:
            return row_seq.pop()
        elif len(col_seq) == 1 and not EMPTY in col_seq:
            return col_seq.pop()
    # check for sequence in diagonal elements & return if found
    if len(diag) == 1 and not EMPTY in diag:
        return diag.pop()
    elif len(cross_diag) == 1 and not EMPTY in cross_diag:
        return cross_diag.pop()
    else:
        return EMPTY


def terminal(board):
    """
    Returns True if game is over, False otherwise.
    """
    board_set = set([tile for row in board for tile in row])
    return not (EMPTY in board_set and winner(board) == None)


def utility(board):
    """
    Returns 1 if X has won the game, -1 if O has won, 0 otherwise.
    """
    Winner = winner(board)
    if Winner == X:
        return 1
    elif Winner == O:
        return -1
    else:
        return 0


def minimax(board):
    """
    Returns the optimal action for the current player on the board.
    """
    if terminal(board):
        return None
    # find which pllayer's turn it is in the board
    turn = player(board) == X
    # find the playable moves init dict to map value to optimised moves
    moves, v2action = actions(board), {}
    # shuffle for randomness
    shuffle(moves)

    minimum_score, maximum_score = -10, 10
    
    def alphabeta(b, alpha, beta, maximizingPlayer):
        """ min_/max_value function with alpha-beta pruning"""

        if terminal(b):
            return utility(b)

        else:
            score, optimizer = (minimum_score, max) if maximizingPlayer else (maximum_score, min)

            possible_moves = actions(b)

            for move in possible_moves:
                child = result(board=b, action=move)

                score = optimizer(score, alphabeta(b=child, alpha=alpha, beta=beta, maximizingPlayer=not maximizingPlayer))

                if maximizingPlayer:
                    alpha = optimizer(alpha, score)
                else:
                    beta = optimizer(beta, score)

                if alpha >= beta:
                    break

            return score
    
    # picks action a in ACTIONS(s) that produces optimal value of MIN/MAX-VALUE(RESULT(s, a))
    for move in moves:
        possible_board = result(board=board, action=move)
        v2action[alphabeta(b=possible_board, maximizingPlayer=not turn, alpha=minimum_score, beta=maximum_score)] = move
    # choose our optimizer function (min or max) based on which player is playing
    optimizer = max if turn else min
    # return the optimised action from dict.
    return v2action[optimizer(v2action)]

File: tictactoe.jl
using LinearAlgebra, Random


const EMPTY = Int8(0)
is_empty = iszero
const X = Int8(1)
const O = Int8(-1)

"""
Returns starting state of the board.
"""
initial_state() = repeat([EMPTY], 3, 3)

"""
Returns player who has the next turn on a board.
"""
player(board) = any(is_empty, board) ? (if (sign∘sum)(board) == X O else X end) : EMPTY

"""
Returns set of all possible actions (i, j) available on the board.
"""
actions(board) = findall(is_empty, board)

"""
Returns the board that results from making move (i, j) on the board.
"""
function result(board, action)
  board_copy = copy(board)
  board_copy[action] = is_empty(board_copy[action]) ? player(board_copy) : board[board_copy]
  
  board_copy
end
result(board, action::Tuple{2}) = result(board, CartesianIndex(action))
result(board, i::T, j::T) where T <: Integer = result(board, CartesianIndex(i, j))

"""
Returns the winner of the game, if there is one.
"""
function winner(board)
  # go through board and find: 
  #   row sequence, col_sequence, diagonal sequence, cross diagonal sequence
  board_len = size(board, 1)
  diag_set, cross_diag_set = Vector{Int8}(undef, 3), Vector{Int8}(undef, 3)
  #   if any of the sequences is of length 3, return the player
  for i in 1:board_len
    diag_set[i], cross_diag_set[i] = board[i, i], board[i, board_len - i + 1]

    (sum(board[i, :]) == 3X || sum(board[:, i]) == 3X) && (return X)
    (sum(board[i, :]) == 3O || sum(board[:, i]) == 3O) && (return O)
  end
  (sum(diag_set) == 3X || sum(cross_diag_set) == 3X) && (return X)
  (sum(diag_set) == 3O || sum(cross_diag_set) == 3O) && (return O)
  #   if no sequences of length 3, return missing
  EMPTY
end

"""
Returns True if game is over, False otherwise.
"""
terminal(board) = winner(board) != EMPTY || EMPTY ∉ board

"""
Returns the optimal action for the current player on the board.
"""
function minimax(board; default_return_move=CartesianIndex(-1, -1))
  terminal(board) && (return default_return_move)
  # find which pllayer's turn it is in the board
  turn = player(board)
  # find the playable moves
  possible_moves = actions(board)
  # shuffle for randomness
  shuffle!(possible_moves)
  # an object to map value to optimised moves
  value_to_move = Vector{Tuple{Int8, CartesianIndex{2}}}(undef, length(possible_moves))
  # picks action a in ACTIONS(s) that produces optimal value of MIN/MAX-VALUE(RESULT(s, a))
  for (i, move) in enumerate(possible_moves)
    possible_board = result(board, move)
    score = alphabeta(possible_board, MINIMUM_SCORE, MAXIMUM_SCORE, turn != X)
    value_to_move[i] = (score, move)
  end
  # choose our optimizer function (min or max) based on which player is playing
  optimizer = turn == X ? argmax : argmin
  # return the optimised action from dict.values
  optimal_value, optimal_move = optimizer(x->x[1], value_to_move)
  optimal_move
end

const MINIMUM_SCORE = Int8(-10)
const MAXIMUM_SCORE = Int8(10)
""" min_/max_value function with alpha-beta pruning"""
function alphabeta(board, α, β, maximizingPlayer)
  terminal(board) && (return winner(board))

  score, optimizer = maximizingPlayer ? (MINIMUM_SCORE, max) : (MAXIMUM_SCORE, min)
  
  possible_moves = actions(board)

  for move in possible_moves
    child = result(board, move)
    
    score = optimizer(score, alphabeta(child, α, β, !maximizingPlayer))

    if maximizingPlayer
      α = optimizer(α, score)
    else
      β = optimizer(β, score)
    end
    
    (α >= β) && (break)
  end
  
  score
end